[user-cuda-streams] Pass streams/events to the graph via lookup table

mlazos opened the pull request:
Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #162905
* #164523
* #165356
* #165212
* #165211
* #164819
* #164522
* #164304
* #162901
* #164507
* #164344
* #164343
* #162903
* #162900
* #163028
* __->__ #162899
* #163027



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

mlazos committed:
[user-cuda-streams] Pass streams/events to the graph via lookup table

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos commented on torch/_dynamo/utils.py:
```diff
@@ -4687,7 +4687,50 @@ def _extract_tensor_dict(t: torch.Tensor) -> dict[str, Any]:
 
+index_to_user_object_weakref: dict[int, weakref.ReferenceType[Any]] = {}
+
+
```
Maybe we could add this as an attribute on the graph

anijain2305 commented on torch/_dynamo/output_graph.py:
```diff
@@ -1413,6 +1415,21 @@ def compile_subgraph(
 
         from .decorators import disable
 
+        if has_user_objects():
```
Add some comments on the index_to_user_object_weakref being a sync point between the pre-graph bytecode and compiled object.

anijain2305 commented on torch/_dynamo/utils.py:
```diff
@@ -4687,7 +4687,50 @@ def _extract_tensor_dict(t: torch.Tensor) -> dict[str, Any]:
 # e.g. torch.cuda.Event is a prime example.
 user_obj_id_to_weakref: dict[int, weakref.ReferenceType[object]] = {}
 
+# Similar to above, except instead of ID, we use a dynamo-generated index
```
Consider moving all of this to a separate file, where we can add these sync points

anijain2305 commented on torch/_dynamo/variables/streams.py:
```diff
@@ -70,11 +70,19 @@ def call_method(
+            from ..guards import GuardBuilder, install_guard
+
+            if self.source:
+                install_guard(self.source.make_guard(GuardBuilder.EQUALS_MATCH))
```
Does EQUAL_MATCH really work? Do streams have __eq__ defined?

anijain2305 reviewed (approved):
Requesting a few changes. Pre-approving, but please consider those requests.

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos committed:
Update on "[user-cuda-streams] Pass streams/events to the graph via lookup table"

Stores streams in a global object look table that maps a dynamo selected index to objects. This index is generated during tracing, and at runtime, a helper function is called from the bytecode to populate this map. 

This differs from the previous implementation that simply mapped IDs to the associated objects. This required specialization on the IDs of the specific objects, while this new approach does not. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
